components/script/dom/htmlimageelement.rs

```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
+        let window = document.window();
+        let task_source = window.dom_manipulation_task_source();
+        let _ = task_source.queue(task, window.upcast());
+    }
+
+    fn set_current_request_url_to_selected_fire_error_loadend(&self, src: DOMString) {
+        struct SetUrlToSelectedTask {
+            img: Trusted<HTMLImageElement>,
+            src: String,
+        }
+        impl Runnable for SetUrlToSelectedTask {
+            fn handler(self: Box<Self>) {
+                // https://html.spec.whatwg.org/multipage/#update-the-image-data
+                // Step 11, substep 5
+                let img = self.img.root();
+                img.current_request.borrow_mut().source_url = Some(self.src.into());
+                img.upcast::<EventTarget>().fire_event(atom!("error"));
+                img.upcast::<EventTarget>().fire_event(atom!("loadend"));
+            }
+        }
+        let runnable = box SetUrlToSelectedTask {
+            img: Trusted::new(self),
+            src: src.into(),
+        };
+        let document = document_from_node(self);
+        let window = document.window();
+        let task = window.dom_manipulation_task_source();
+        let _ = task.queue(runnable, window.upcast());
+    }
+
+    /// Step 12 of html.spec.whatwg.org/multipage/#update-the-image-data
+    fn fetch_image(&self, url: &ServoUrl, src: &DOMString) {
+        let mut current_request = self.current_request.borrow_mut();
+        let mut pending_request = self.pending_request.borrow_mut();
+        if let Some(pending_url) = pending_request.parsed_url.clone() {
+            // Step 12.1
+            if pending_url == *url {
+                return
+            }
+            // Step 12.2, 12.3 abort pending request
+            pending_request.image = None;
+            pending_request.parsed_url = None;
+            LoadBlocker::terminate(&mut pending_request.blocker);
+        }
+        // step 12.4, create a new "image_request"
+        let mut image_request = match (current_request.parsed_url.clone(), current_request.state) {
+            (Some(parsed_url), State::PartiallyAvailable) => {
+                // Step 12.2
+                if parsed_url == *url {
+                    // TODO: queue a task to restart animation, if restart-animation is set
+                    return
+                }
+                pending_request
+            },
+            (_, State::Broken) | (_, State::Unavailable) => {
+                // Step 12.5
+                LoadBlocker::terminate(&mut current_request.blocker);
+                current_request
+            },
+            (_, _) => {
+                // step 12.6
+                pending_request
+            },
+        };
+        image_request.parsed_url = Some(url.clone());
+        image_request.source_url = Some(src.clone());
+        image_request.image = None;
+        image_request.metadata = None;
+        let document = document_from_node(self);
```

Comment:
the actual fetch request is initialized a bit futher down the road, [here](https://github.com/servo/servo/blob/master/components/script/dom/htmlimageelement.rs#L261), after a cache check. I'm not sure if we need to change this, even it is indeed at this stage(step 12.6) that the spec speaks of "creating a potential-CORS request"...